### PR TITLE
[refactoring] Create common interface for CMutableTransaction and CTransaction with static polymorphism

### PR DESCRIPTION
**Pull request by lucash-dev**
This PR creates a common interface for `CMutableTransaction` and `CTransaction` as the template class `CBaseTransaction<T>`. This base class uses the CRTP (curiously recurring template pattern) in order to achieve static polymorphism. This way this change shouldn't have any effect on behaviour or runtime performance (though compilation performance is of course worse).

Rationale:
 - Creates a common, explicit interface type for `CTransaction` and `CMutableTransaction`.
 - Makes explicit common elements between two classes.
 - Functions that accept have a better option than accepting an unrestricted typename T.
 - Better than proliferating custom `static_assert` or `enable_if` with different, obscure error messages when the template is applied to the wrong types.

PS: This PR shouldn't affect performance or behaviour, but please review thoroughly as it touches consensus-sensitive code.

### CODE REVIEW

**Comment by gmaxwell** [comment]
The PR description doesn't seem to state why this would be a benefit for users of the system, only why it wouldn't be a disaster.   As a result, this does not appear to justify the effort to review it for safety.

**Comment by lucash-dev** [comment]
@gmaxwell I understand your concerns. This PR is about code maintainability and extensibility, which is beneficial to the user in the long run (if it’s a good refactoring).

But It’s hard for me to judge the effort required of others to review it

File: src/primitives/transaction.h [hunk 1]
```diff
@@ -388,15 +444,6 @@ struct CMutableTransaction
      */
     uint256 GetHash() const;
```
Comment (practicalswift) [inline src/primitives/transaction.h]: Nit: Remove blank line in end of code block :-)

**Comment by lucash-dev** [comment]
@gmaxwell after a bit more thought, I actually agree with you. Closing the PR. 
Thank you very much for taking the time to look at it and sharing your insights.

**Pull request closed without merging**